### ISSUE TITLE
Adding a helper function to compute snapshots used in HODMD?

### ISSUE THREAD

**Issue by adam2392** [mathLab/PyDMD#143]
# Main issue
I was wondering if it would be okay with developers to add in a public function to HoDMD to recompute the snapshots?

E.g.

```Python
def get_snapshots():
        snaps = np.concatenate(
            [
                self._snapshots[:, i:self._snapshots.shape[1] - self.d + i + 1]
                for i in range(self.d)
            ],
            axis=0)
      return snaps
```

The reason is, I am trying to rescale the DMD modes computed and some of the algorithms for rescaling want to do underlying computations on the original snapshots ran through HoDMD. 

# Issue about allowing different scaling strategies
With that being said a small side issue I was having is how to rescale the modes. For example, in this paper on page 5, authors have scaled the modes by the singular values of the corresponding SVD, so the eigenvectors are `W = \Sigma^{1/2} \hat{W}` instead of `\hat{W}`, which is then used to compute the DMD modes. 

Would it be desirable to either i) add alternative scaling strategies on the DMD modes or ii) separate out the steps inside `fit()` a bit more such that, if someone wanted to alter say in between steps, they can start off there and call some private/public functions?

**Comment by ndem0** [comment]
Dear @adam2392,
yes, we can easily add a function in the `HODMD` class for arranging the snapshots matrix.

But, actually I didn't get why this should help you in rescaling the DMD modes. Do you need a public function that returns the 'higher-order' snapshots matrix or a way to post-process the DMD modes?

**Comment by adam2392** [comment]
Hi @ndem0 thanks for the response. Since posting this, I found a way to rescale the DMD modes by its corresponding singular values from the input data matrix according to the paper I linked.

```Python
def rescale_modes(snapshots, atilde, dmd):
    """Helper function to rescale and recompute DMD modes.

    Re-scales the DMD modes by the singular values from
    the data matrix SVD.

    Parameters
    ----------
    snapshots : np.ndarray
        The original snapshots matrix.
    atilde : np.ndarray
        The fitted rank-truncated linear operator from DMD.
    dmd : instance of pydmd.DMDBase
        ``dmd`` should already have been fitted.

    Returns
    -------
    modes : np.ndarray
        The re-scaled DMD modes.
    """
    Y = snapshots[:, 1:]

    # compute the SVD
    U, s, V = dmd._compute_svd(snapshots[:, :-1], dmd.svd_rank)

    # compute Ahat from Atilde
    sigma_inv_sqrt = np.diag(np.power(s.copy(), -0.5))
    sigma_sqrt = np.diag(np.power(s.copy(), 0.5))
    Ahat = sigma_inv_sqrt.dot(atilde).dot(sigma_sqrt)

    # compute eigenvector decomposition of Atilde
    lowrank_eigenvalues, lowrank_eigenvectors = np.linalg.eig(Ahat)

    # check norms
    #     print(np.linalg.norm(lowrank_eigenvectors, axis=0))

    # rescale eigenvectors by singular values
    lowrank_eigenvectors = sigma_sqrt.dot(lowrank_eigenvectors)

    #     print(Ahat.shape, lowrank_eigenvectors.shape)
    #     print(np.linalg.norm(lowrank_eigenvectors, axis=0))

    # compute DMD modes again
    modes = (Y.dot(V).dot(np.diag(np.reciprocal(s)))).dot(lowrank_eigenvectors)
    modes = modes[:dmd._snapshots.shape[0], :]
    return modes
```

I suppose what might be useful is a public function called `get_time_lagged_snapshots()` that returns the following array:

```
# snapshots is the input data matrix; d is the order in HODMD
snaps = np.concatenate(
    [
        snapshots[:, i:snapshots.shape[1] - d + i + 1]
        for i in range(d)
    ],
    axis=0)
```
?

In an ideal world, there would be some way of specifying how to "re-scale" the modes in different ways, but I can see how that might become hard to manage and abstract into an API.

**Comment by adam2392** [comment]
If you think that the above is useful, I can add the `get_time_lagged_snapshots()` function to HoDMD in a PR.

**Comment by ndem0** [comment]
I still be a bit confused. I would separate the two things you are talking about, just to avoid misunderstandings:
- rescaling of the modes: from the code you posted, I see that both the reduced operator `Atilde` and the modes are rescaled, not only the modes. Am I right? Then probably the best option is to add a function to `DMDBase` class to manage the scaling in all the DMD versions implemented within PyDMD, not just HODMD. Meanwhile may I ask you the paper where this scaling is proposed?
- the function for the *lagged* snapshots is ok!

**Comment by adam2392** [comment]
> * rescaling of the modes: from the code you posted, I see that both the reduced operator `Atilde` and the modes are rescaled, not only the modes. Am I right? Then probably the best option is to add a function to `DMDBase` class to manage the scaling in all the DMD versions implemented within PyDMD, not just HODMD. Meanwhile may I ask you the paper where this scaling is proposed?

Yes this particular rescaling scales the modes wrt singular values from the data snapshots matrix. I can imagine there would be necessary times to rescale the modes. In this case, though, it would be an alternative calculation as essentially the entire inner-workings of `fit()` is redone (SVD, rescaled Atilde computation, eigenvector decomposition and then rescaled modes).

Yes, sorry I thought I had linked the paper: https://arxiv.org/pdf/1409.5496.pdf :p.

> * the function for the _lagged_ snapshots is ok!

Okay will add a PR for this!

**Comment by ndem0** [comment]
Dear @adam2392 I've read the paper you shared: the new algorithm just differs from the standard by a couple of matrix multiplications.
This let me think that we can simply add a class parameter (e.g. `rescale_mode`) to DMD to control such behaviour:
- 'auto' for automating rescaling using the singular values (as done in the paper)
- None for no rescaling. Default behaviour.
- a numpy.array to give to the final user the possibility to test different rescaling. Of course the number of components in this array should matches the number of modes.

Can be a valid option for you? I prefer this option instead of creating a new class (for example `RescaledDMD`) to give the user more flexibility

**Comment by adam2392** [comment]
> Dear @adam2392 I've read the paper you shared: the new algorithm just differs from the standard by a couple of matrix multiplications.
> This let me think that we can simply add a class parameter (e.g. `rescale_mode`) to DMD to control such behaviour:
> 
> * 'auto' for automating rescaling using the singular values (as done in the paper)
> * None for no rescaling. Default behaviour.
> * a numpy.array to give to the final user the possibility to test different rescaling. Of course the number of components in this array should matches the number of modes.
> 
> Can be a valid option for you? I prefer this option instead of creating a new class (for example `RescaledDMD`) to give the user more flexibility

Yes this sounds a lot better then adding a new class!

Regarding the `numpy.array`, what do you mean exactly? I'm a bit confused on this one. Is this meant to scale the modes directly via `modes.dot(numpy.array passed in)`, or is it an array that scales a different step in the DMD algorithm?

Everything else is fairly clear, and I'm happy to make a PR sometime later when I have some time.

**Comment by ndem0** [comment]
Yes, my idea was to use the passed `numpy.array` to scale the modes and the operator using something like `modes.dot(passed_array)`. Using the argument 'auto' this array contains the singular values (as in the paper) but it allows also for other rescaling (e.g. the square of singular values). I really don't know if is something that can useful playing with the rescaling, the aim is just to not limit the user to some _fixed_ behaviour.
From the code perspective, this modification is straightforward.

**Comment by ndem0** [comment]
Fixed in #151
